### ISSUE TITLE
Bridge notifications retained messages duplicated/not correct on remote broker. 

### ISSUE THREAD

**Issue by albe1066** [eclipse/mosquitto#1510]
mosquitto 1.6.7 on remote broker - Bridge notifications are duplicated and last message incorrect on remote broker. On local broker sometimes duplicates but the state of connection is correct. We do not see this behavior with mosquitto 1.4.14 on client and 1.6.7 on remote broker. 

Restart mosquitto on local device 1.6.7
local shows
no disconnect when mosquitto stopped --  1.4.14 sends disconnect to local
mosquitto connects to remote broker
device/o/s/s/c/h 0   - this is the missing disconnect but sent after connect 1.4.14 sends when disconnected 
device/o/s/s/c/h 1  -- correct connect

remote broker shows
device/o/s/s/c/h 0 -- Disconnect 
then when reconnects 
device/o/s/s/c/h 1
device/o/s/s/c/h 1
device/o/s/s/c/h 0
bridge is connected but notification shows on remote cloud as disconnected. 

Bridge file
```
connection localtocloud
start_type automatic
cleansession false
max_inflight_messages 0
max_queued_messages 200
clientid localtocloud
address  remotebrokeraddress:port

topic  device/i/# in 0
topic  device/o/# out 2

try_private true
notifications true
"# notifications_local_only false -- using default"
notification_topic device/o/s/s/c/h
bridge_protocol_version mqttv31

remote_username <remoteusername>
remote_password <remotepassword>

bridge_cafile <pathtocert/cert>
tls_version tlsv1.2
```
Log from remote mosquitto server 
The following is off a server using mosquitto_auth_plugin - same behavior on server using mosquitto password and acl files

```
mosquitto_1      | 1574709310: |-- mosquitto_auth_acl_check(..., devicetocloud, localtocloud, device/o/s/s/c/h, 2)
mosquitto_1      | 1574709310: Socket error on client devicetocloud, disconnecting.
mosquitto_1      | 1574709310: |--  Cached  [67CD462217DABA31CB7C5B9A60FF7AA35B2D68FA] for (devicetocloud,localtocloud,2)
mosquitto_1      | 1574709356: New bridge connected from ipaddress as devicetocloud (p1, c0, k60, u'localtocloud').
mosquitto_1      | 1574709356: Sending CONNACK to devicetocloud (0, 0)
mosquitto_1      | 1574709356: |-- mosquitto_auth_acl_check(..., devicetocloud, localtocloud, device/o/s/s/c/h, 2)
mosquitto_1      | 1574709356: Received PUBLISH from devicetocloud (d0, q1, r1, m9, 'device/o/s/s/c/h', ... (1 bytes))
mosquitto_1      | 1574709356: Sending PUBACK to devicetocloud (m9, rc0)
mosquitto_1      | 1574709356: Received SUBSCRIBE from devicetocloud
mosquitto_1      | 1574709356: devicetocloud 0 device/i/#
mosquitto_1      | 1574709356: Sending SUBACK to devicetocloud
mosquitto_1      | 1574709356: Received PUBLISH from devicetocloud (d0, q1, r1, m11, 'device/o/s/s/c/h', ... (1 bytes))
mosquitto_1      | 1574709356: Sending PUBACK to devicetocloud (m11, rc0)
mosquitto_1      | 1574709356: |-- mosquitto_auth_acl_check(..., devicetocloud, localtocloud, device/o/s/s/c/h, 2)
mosquitto_1      | 1574709356: |-- mosquitto_auth_acl_check(..., devicetocloud, localtocloud, device/o/s/s/c/h, 2)
mosquitto_1      | 1574709356: Received PUBLISH from devicetocloud (d0, q1, r1, m8, 'device/o/s/s/c/h', ... (1 bytes))
mosquitto_1      | 1574709356: Sending PUBACK to devicetocloud (m8, rc0)
```

Any idea why 1.6.7 has different bridge notification behavior?
Is this a defect?

**Comment by albe1066** [comment]
I have no idea why some of the above is bold and larger text.

**Comment by ralight** [comment]
I think what is happening here is a bit of an odd case that comes about because you are bridging the topic where the notifications are being sent, so they are both duplicated and then queued up for later delivery when the bridge is offline (because you have clean session set to false and are using QoS>0).

Do you see the same behaviour if you change the notification topic to a topic that you aren't bridging?

**Comment by albe1066** [comment]
Hi I can test what you suggest. Here is something I discovered during test beginning with 1.5.5 to 1.6.7 client local broker bridging 1.6.7 remote broker. Given the following 
max_inflight_messages 0
max_queued_messages 1
When bridge connects (reconnect wan)  remote broker shows connect notification followed by 20 messages - first messages sent to local after disconnect - followed by another connect notification then all retained messages. All messages on local bridged to remote are retained. If inflight and max_queued are set to different values I do not see all retained messages sent to remote broker.

**Comment by ralight** [comment]
Thanks for the extra information - can I ask though that you try to give precise details to reproduce the behaviour that you are seeing. At the moment your latest description isn't enough for me. I need to know exactly what you did, otherwise I can't reproduce it.

**Comment by albe1066** [comment]
Roger I can remove application specifics from test script to reproduce behavior and upload with exactly what I am doing to disconnect and reconnect if useful?

**Comment by ralight** [comment]
If you have a test script that would be even better than a description.

**Comment by albe1066** [comment]
Changed topic for notification so it is not the same as bridge out. This change stopped duplicate messages. I changed max_queued_messages to 20 - when bridge connects all 50 retained messages are sent. Changed max_inflight_messages to 1 with max_queued_messages 20 and received all retained messages on reconnect.  I did not see any difference in order or number of messages sent to remote broker when network reconnects when changing the above settings. 

The change of notification topic solved the duplicate message issue. We want only retained messages sent when local connects to remote. We were successful getting this behavior.  I will close this ticket. 

If you would like the script used let me know..
